--- a/autoscaler.py
+++ b/autoscaler.py
@@ -1,3 +1,6 @@
+# From https://github.com/ray-project/ray/blob/ray-2.9.3/python/ray/autoscaler/_private/autoscaler.py
+# Sky patch changes:
+#  - enable upscaling_speed to be 0.0
 import copy
 import logging
 import math
@@ -1071,7 +1074,7 @@
             upscaling_speed = self.config.get("upscaling_speed")
             aggressive = self.config.get("autoscaling_mode") == "aggressive"
             target_utilization_fraction = self.config.get("target_utilization_fraction")
-            if upscaling_speed:
+            if upscaling_speed is not None: # NOTE(sky): enable 0.0
                 upscaling_speed = float(upscaling_speed)
             # TODO(ameer): consider adding (if users ask) an option of
             # initial_upscaling_num_workers.
