--- a/log_monitor.py
+++ b/log_monitor.py
@@ -1,3 +1,7 @@
+# Original file https://github.com/ray-project/ray/blob/ray-2.9.3/python/ray/_private/log_monitor.py
+# Fixed the problem for progress bar, as the latest version does not preserve \r for progress bar.
+# We change the newline handling back to https://github.com/ray-project/ray/blob/ray-1.10.0/python/ray/_private/log_monitor.py#L299-L300
+
 import argparse
 import errno
 import glob
@@ -374,7 +378,8 @@
                     next_line = next_line.decode("utf-8", "replace")
                     if next_line == "":
                         break
-                    next_line = next_line.rstrip("\r\n")
+                    if next_line.endswith("\n"):
+                        next_line = next_line[:-1]
 
                     if next_line.startswith(ray_constants.LOG_PREFIX_ACTOR_NAME):
                         flush()  # Possible change of task/actor name.
