--- a/resource_demand_scheduler.py
+++ b/resource_demand_scheduler.py
@@ -1,3 +1,8 @@
+# From https://github.com/ray-project/ray/blob/ray-2.9.3/python/ray/autoscaler/_private/resource_demand_scheduler.py
+# Sky patch changes:
+#  - no new nodes are allowed to be launched launched when the upscaling_speed is 0
+#  - comment out "assert not unfulfilled": this seems a buggy assert
+
 """Implements multi-node-type autoscaling.
 
 This file implements an autoscaling algorithm that is aware of multiple node
@@ -448,7 +453,10 @@
                 + placement_group_nodes.get(node_type, 0),
             )
 
-            if upper_bound > 0:
+            # NOTE(sky): do not autoscale when upsclaing speed is 0.
+            if self.upscaling_speed == 0:
+                upper_bound = 0
+            if upper_bound >= 0:
                 updated_nodes_to_launch[node_type] = min(
                     upper_bound, to_launch[node_type]
                 )
@@ -592,7 +600,7 @@
             unfulfilled, including_reserved = get_bin_pack_residual(
                 new_node_resources, unfulfilled, strict_spread=True
             )
-            assert not unfulfilled
+            # assert not unfulfilled  # NOTE(sky): buggy assert.
             node_resources += including_reserved
         return to_add, node_resources, node_type_counts
 
