--- a/worker.py
+++ b/worker.py
@@ -1,3 +1,7 @@
+# Adapted from https://github.com/ray-project/ray/blob/ray-2.56.1/python/ray/_private/worker.py
+# Fixed the problem in ray's issue https://github.com/ray-project/ray/issues/9233
+# Tracked in PR https://github.com/ray-project/ray/pull/21977/files.
+
 import atexit
 import dataclasses
 import faulthandler
@@ -2347,6 +2351,13 @@
         pid = data.get("pid")
         lines = data.get("lines", [])
 
+    def end_for(line: str) -> str:
+        if sys.platform == "win32":
+            return "\n"
+        if line.endswith("\r"):
+            return ""
+        return "\n"
+
     ip = data.get("ip")
     ip_prefix = "" if ip == data.get("localhost") else f", ip={ip}"
     for line in lines:
@@ -2366,12 +2377,14 @@
                 print(
                     f"{message_for(data, line)}",
                     file=print_file,
+                    end=end_for(line),
                 )
             else:
                 print(
                     f"{color_pre}({prefix_for(data)}{pid}{ip_prefix}){color_post} "
                     f"{message_for(data, line)}",
                     file=print_file,
+                    end=end_for(line),
                 )
     if ray_constants.RAY_FLUSH_DRIVER_LOGS:
         if hasattr(print_file, "flush"):
