--- a/worker.py
+++ b/worker.py
@@ -1,3 +1,7 @@
+# Adapted from https://github.com/ray-project/ray/blob/ray-2.9.3/python/ray/_private/worker.py
+# Fixed the problem in ray's issue https://github.com/ray-project/ray/issues/9233
+# Tracked in PR https://github.com/ray-project/ray/pull/21977/files.
+
 import atexit
 import faulthandler
 import functools
@@ -2020,6 +2024,14 @@
         pid = data.get("pid")
         lines = data.get("lines", [])
 
+    def end_for(line: str) -> str:
+        if sys.platform == "win32":
+            return "\n"
+        if line.endswith("\r"):
+            return ""
+        return "\n"
+
+
     if data.get("ip") == data.get("localhost"):
         for line in lines:
             if RAY_TQDM_MAGIC in line:
@@ -2035,6 +2047,7 @@
                         message_for(data, line),
                     ),
                     file=print_file,
+                    end=end_for(line),
                 )
     else:
         for line in lines:
@@ -2052,6 +2065,7 @@
                         message_for(data, line),
                     ),
                     file=print_file,
+                    end=end_for(line),
                 )
     # Restore once at end of batch to avoid excess hiding/unhiding of tqdm.
     restore_tqdm()
